# Assignment-1

## Simple Model

In [414]:
import sklearn.datasets
import numpy as np
import pandas as pd
from sklearn.model_selection import train_test_split

In [415]:
cancer_ds=sklearn.datasets.load_breast_cancer()
X=cancer_ds.data                     #Separating data and target also Converting into ndarray
Y=cancer_ds.target
print(X.shape, Y.shape)

(569, 30) (569,)


In [416]:
data = pd.DataFrame(cancer_ds.data, columns=cancer_ds.feature_names)
data['Class'] = cancer_ds.target         #appending new column "Class" to the data with its values
data.head()

mean radius  mean texture  mean perimeter  mean area  mean smoothness  \
0        17.99         10.38          122.80     1001.0          0.11840   
1        20.57         17.77          132.90     1326.0          0.08474   
2        19.69         21.25          130.00     1203.0          0.10960   
3        11.42         20.38           77.58      386.1          0.14250   
4        20.29         14.34          135.10     1297.0          0.10030   

   mean compactness  mean concavity  mean concave points  mean symmetry  \
0           0.27760          0.3001              0.14710         0.2419   
1           0.07864          0.0869              0.07017         0.1812   
2           0.15990          0.1974              0.12790         0.2069   
3           0.28390          0.2414              0.10520         0.2597   
4           0.13280          0.1980              0.10430         0.1809   

   mean fractal dimension  ...  worst texture  worst perimeter  worst area  \
0                 0.07871  ...          17.33           184.60      2019.0   
1                 0.05667  ...          23.41           158.80      1956.0   
2                 0.05999  ...          25.53           152.50      1709.0   
3                 0.09744  ...          26.50            98.87       567.7   
4                 0.05883  ...          16.67           152.20      1575.0   

   worst smoothness  worst compactness  worst concavity  worst concave points  \
0            0.1622             0.6656           0.7119                0.2654   
1            0.1238             0.1866           0.2416                0.1860   
2            0.1444             0.4245           0.4504                0.2430   
3            0.2098             0.8663           0.6869                0.2575   
4            0.1374             0.2050           0.4000                0.1625   

   worst symmetry  worst fractal dimension  Class  
0          0.4601                  0.11890      0  
1          0.2750                  0.08902      0  
2          0.3613                  0.08758      0  
3          0.6638                  0.17300      0  
4          0.2364                  0.07678      0  

[5 rows x 31 columns]

In [417]:
data.describe()

mean radius  mean texture  mean perimeter    mean area  \
count   569.000000    569.000000      569.000000   569.000000   
mean     14.127292     19.289649       91.969033   654.889104   
std       3.524049      4.301036       24.298981   351.914129   
min       6.981000      9.710000       43.790000   143.500000   
25%      11.700000     16.170000       75.170000   420.300000   
50%      13.370000     18.840000       86.240000   551.100000   
75%      15.780000     21.800000      104.100000   782.700000   
max      28.110000     39.280000      188.500000  2501.000000   

       mean smoothness  mean compactness  mean concavity  mean concave points  \
count       569.000000        569.000000      569.000000           569.000000   
mean          0.096360          0.104341        0.088799             0.048919   
std           0.014064          0.052813        0.079720             0.038803   
min           0.052630          0.019380        0.000000             0.000000   
25%           0.086370          0.064920        0.029560             0.020310   
50%           0.095870          0.092630        0.061540             0.033500   
75%           0.105300          0.130400        0.130700             0.074000   
max           0.163400          0.345400        0.426800             0.201200   

       mean symmetry  mean fractal dimension  ...  worst texture  \
count     569.000000              569.000000  ...     569.000000   
mean        0.181162                0.062798  ...      25.677223   
std         0.027414                0.007060  ...       6.146258   
min         0.106000                0.049960  ...      12.020000   
25%         0.161900                0.057700  ...      21.080000   
50%         0.179200                0.061540  ...      25.410000   
75%         0.195700                0.066120  ...      29.720000   
max         0.304000                0.097440  ...      49.540000   

       worst perimeter   worst area  worst smoothness  worst compactness  \
count       569.000000   569.000000        569.000000         569.000000   
mean        107.261213   880.583128          0.132369           0.254265   
std          33.602542   569.356993          0.022832           0.157336   
min          50.410000   185.200000          0.071170           0.027290   
25%          84.110000   515.300000          0.116600           0.147200   
50%          97.660000   686.500000          0.131300           0.211900   
75%         125.400000  1084.000000          0.146000           0.339100   
max         251.200000  4254.000000          0.222600           1.058000   

       worst concavity  worst concave points  worst symmetry  \
count       569.000000            569.000000      569.000000   
mean          0.272188              0.114606        0.290076   
std           0.208624              0.065732        0.061867   
min           0.000000              0.000000        0.156500   
25%           0.114500              0.064930        0.250400   
50%           0.226700              0.099930        0.282200   
75%           0.382900              0.161400        0.317900   
max           1.252000              0.291000        0.663800   

       worst fractal dimension       Class  
count               569.000000  569.000000  
mean                  0.083946    0.627417  
std                   0.018061    0.483918  
min                   0.055040    0.000000  
25%                   0.071460    0.000000  
50%                   0.080040    1.000000  
75%                   0.092080    1.000000  
max                   0.207500    1.000000  

[8 rows x 31 columns]

In [418]:
print(data['Class'].value_counts()) #Counting the Binary values

1    357
0    212
Name: Class, dtype: int64


In [419]:
print(cancer_ds.target_names)

['malignant' 'benign']


In [420]:
data.groupby('Class').mean()

mean radius  mean texture  mean perimeter   mean area  mean smoothness  \
Class                                                                           
0        17.462830     21.604906      115.365377  978.376415         0.102898   
1        12.146524     17.914762       78.075406  462.790196         0.092478   

       mean compactness  mean concavity  mean concave points  mean symmetry  \
Class                                                                         
0              0.145188        0.160775             0.087990       0.192909   
1              0.080085        0.046058             0.025717       0.174186   

       mean fractal dimension  ...  worst radius  worst texture  \
Class                          ...                                
0                    0.062680  ...     21.134811      29.318208   
1                    0.062867  ...     13.379801      23.515070   

       worst perimeter   worst area  worst smoothness  worst compactness  \
Class                                                                      
0           141.370330  1422.286321          0.144845           0.374824   
1            87.005938   558.899440          0.124959           0.182673   

       worst concavity  worst concave points  worst symmetry  \
Class                                                          
0             0.450606              0.182237        0.323468   
1             0.166238              0.074444        0.270246   

       worst fractal dimension  
Class                           
0                     0.091530  
1                     0.079442  

[2 rows x 30 columns]

In [421]:
X = data.drop('Class', axis=1)             #without column class
Y= data['Class']

## random_state attribute of train_test_split function
The same data is gone under train and test we can see that in the below demonstration taking only 0.2% data as test data so that we can easily compare between the data

In [422]:
print("The Output format is\nIndex no.  Label")
for _ in range(4):
    X_train, X_test, Y_train, Y_test = train_test_split(X,Y,test_size=0.002, random_state=1)#0.2% Testdata
    print(Y_test)

The Output format is
Index no.  Label
421    1
47     0
Name: Class, dtype: int32
421    1
47     0
Name: Class, dtype: int32
421    1
47     0
Name: Class, dtype: int32
421    1
47     0
Name: Class, dtype: int32


## stratify attribute of train_test_split function
The distribution of the data between training and testing is same such that its mean, SD etc are same

In [423]:
for i in range(5):
    X_train, X_test, Y_train, Y_test = train_test_split(X,Y, test_size = 0.2,stratify = Y)
    print(Y_test.mean())

print("*"*30)

for i in range(5): 
    X_train, X_test, Y_train, Y_test = train_test_split(X,Y, test_size = 0.2)
    print(Y_test.mean())

0.631578947368421
0.631578947368421
0.631578947368421
0.631578947368421
0.631578947368421
******************************
0.6491228070175439
0.6929824561403509
0.6666666666666666
0.631578947368421
0.5877192982456141


In [424]:
#Binarization of Input

X_train, X_test, Y_train, Y_test = train_test_split(X,Y, test_size = 0.25,stratify = Y)    #25% test data  
                                                                                #uniform distribution of train and test data

X_binarised_train = X_train.apply(pd.cut, bins=2, labels = [1,0])
X_binarised_test = X_test.apply(pd.cut, bins=2, labels = [1,0])
type(X_binarised_test)

X_binarised_train=X_binarised_train.values
X_binarised_test=X_binarised_test.values

type(X_binarised_test)

numpy.ndarray

# Prediction model using loop

## On Train data

In [425]:
b=0
accurate_rows = 0
max_accurate_rows=0
b_optimum=0
print("For Train data:\n")
print("OUTPUT FORMAT iS \n{b ,Accurate_Rows ,Accuracy }")
for b in range(1,31):
    y_pred=0
    accurate_rows=0
    
    for x,y in zip(X_binarised_train, Y_train):
        y_pred = np.sum(x)>=b       #if summation of x(i) for all features is greater than threshold "b", then predict 1 else 0
        accurate_rows += (y==y_pred)              #Total number of accurate rows 
        accuracy=accurate_rows/X_binarised_train.shape[0]             
    print("  {:02d}    {}    {:.2f}".format(b,accurate_rows, accuracy))
    
    if(max_accurate_rows<accurate_rows):             #if max value less than current then update the values
        max_accurate_rows=accurate_rows
        b_optimum=b
    
        
print("b_optimum=",b_optimum)
print("max_accuracy=",max_accurate_rows/X_binarised_train.shape[0]*100)

For Train data:

OUTPUT FORMAT iS 
{b ,Accurate_Rows ,Accuracy }
  01    267    0.63
  02    267    0.63
  03    267    0.63
  04    267    0.63
  05    267    0.63
  06    267    0.63
  07    267    0.63
  08    267    0.63
  09    267    0.63
  10    267    0.63
  11    267    0.63
  12    267    0.63
  13    270    0.63
  14    270    0.63
  15    271    0.64
  16    275    0.65
  17    278    0.65
  18    282    0.66
  19    283    0.66
  20    290    0.68
  21    293    0.69
  22    301    0.71
  23    313    0.73
  24    325    0.76
  25    344    0.81
  26    356    0.84
  27    363    0.85
  28    362    0.85
  29    348    0.82
  30    309    0.73
b_optimum= 27
max_accuracy= 85.2112676056338


## On test data

In [426]:
test_accurate_rows = 0
test_max_accurate_rows=0
test_b_optimum=0

print("For Test data:\n")
print("OUTPUT FORMAT iS \n{b ,Accurate_Rows ,Accuracy }")
for b in range(1,31):
    y_pred=0
    test_accurate_rows=0
    
    for x,y in zip(X_binarised_test, Y_test):
        y_pred = np.sum(x)>=b
        test_accurate_rows += (y==y_pred)

    test_Accuracy=test_accurate_rows/X_binarised_test.shape[0]
    if(test_max_accurate_rows<test_accurate_rows):
        test_max_accurate_rows=test_accurate_rows
        test_b_optimum=b
    print("  {:02d}    {}    {:.2f}".format(b,test_accurate_rows, test_Accuracy))

print("test_b_optimum=",test_b_optimum)
print("test_max_accuracy=",test_max_accurate_rows/X_binarised_test.shape[0]*100)

For Test data:

OUTPUT FORMAT iS 
{b ,Accurate_Rows ,Accuracy }
  01    90    0.63
  02    90    0.63
  03    90    0.63
  04    90    0.63
  05    90    0.63
  06    90    0.63
  07    90    0.63
  08    90    0.63
  09    90    0.63
  10    91    0.64
  11    91    0.64
  12    91    0.64
  13    92    0.64
  14    93    0.65
  15    96    0.67
  16    96    0.67
  17    98    0.69
  18    102    0.71
  19    104    0.73
  20    106    0.74
  21    110    0.77
  22    113    0.79
  23    116    0.81
  24    121    0.85
  25    122    0.85
  26    123    0.86
  27    123    0.86
  28    121    0.85
  29    106    0.74
  30    90    0.63
test_b_optimum= 26
test_max_accuracy= 86.01398601398601



# Vectorised Implementation
In the above i have implemented the model using For loops but as we know looping is costly in terms of computation so i will also implement it using vectors.
We can eliminate the inner for loop.

## Train data

In [427]:
max_accurate_rows=0
print("For Train data:\n")
print("OUTPUT FORMAT iS \n{b ,Accurate_Rows ,Accuracy }")
for b in range(1,31):
    
    Sum= np.sum(X_binarised_train, axis=1)   #Summation of x(i) values for all features

    y_pred =  (Sum>= b)                      # prediction of y_pred values using threshold = b

    if y_pred.dtype==bool:             #Converting boolean values into integers
        y_pred= y_pred.astype(int)
    
    accurate_rows = np.count_nonzero(y_pred == Y_train)     #counting number of accurately predicted values
    accuracy = np.mean(y_pred == Y_train)                   #accuracy
    
    print("  {:02d}    {}    {:.2f}".format(b,accurate_rows, accuracy))
    
    if(max_accurate_rows<accurate_rows):
        max_accurate_rows=accurate_rows
        b_optimum=b
        
print("b_optimum=",b_optimum)
print("max_accuracy=",max_accurate_rows/X_binarised_train.shape[0]*100)

For Train data:

OUTPUT FORMAT iS 
{b ,Accurate_Rows ,Accuracy }
  01    267    0.63
  02    267    0.63
  03    267    0.63
  04    267    0.63
  05    267    0.63
  06    267    0.63
  07    267    0.63
  08    267    0.63
  09    267    0.63
  10    267    0.63
  11    267    0.63
  12    267    0.63
  13    270    0.63
  14    270    0.63
  15    271    0.64
  16    275    0.65
  17    278    0.65
  18    282    0.66
  19    283    0.66
  20    290    0.68
  21    293    0.69
  22    301    0.71
  23    313    0.73
  24    325    0.76
  25    344    0.81
  26    356    0.84
  27    363    0.85
  28    362    0.85
  29    348    0.82
  30    309    0.73
b_optimum= 27
max_accuracy= 85.2112676056338


## Test Data

In [428]:
max_accurate_rows=0
print("For Test data:\n")
print("OUTPUT FORMAT iS \n{b ,Accurate_Rows ,Accuracy }")
for b in range(1,31):
    
    Sum= np.sum(X_binarised_test, axis=1)   #Summation of x(i) values for all features

    y_pred =  (Sum>= b)                      # prediction of y_pred values using threshold = b

    if y_pred.dtype==bool:             #Converting boolean values into integers
        y_pred= y_pred.astype(int)
    
    accurate_rows = np.count_nonzero(y_pred == Y_test)
    accuracy = np.mean(y_pred == Y_test)
    
    print("  {:02d}    {}    {:.2f}".format(b,accurate_rows, accuracy))
    
    if(max_accurate_rows<accurate_rows):
        max_accurate_rows=accurate_rows
        b_optimum=b
        
print("b_optimum=",b_optimum)
print("max_accuracy=",max_accurate_rows/X_binarised_test.shape[0]*100)

For Test data:

OUTPUT FORMAT iS 
{b ,Accurate_Rows ,Accuracy }
  01    90    0.63
  02    90    0.63
  03    90    0.63
  04    90    0.63
  05    90    0.63
  06    90    0.63
  07    90    0.63
  08    90    0.63
  09    90    0.63
  10    91    0.64
  11    91    0.64
  12    91    0.64
  13    92    0.64
  14    93    0.65
  15    96    0.67
  16    96    0.67
  17    98    0.69
  18    102    0.71
  19    104    0.73
  20    106    0.74
  21    110    0.77
  22    113    0.79
  23    116    0.81
  24    121    0.85
  25    122    0.85
  26    123    0.86
  27    123    0.86
  28    121    0.85
  29    106    0.74
  30    90    0.63
b_optimum= 26
max_accuracy= 86.01398601398601


In [429]:
from sklearn import metrics
Sum= np.sum(X_binarised_test, axis=1)   #Summation of x(i) values for all features
y_pred =  (Sum>= 28)
cm = metrics.confusion_matrix(Y_test, y_pred)  #false-false, true-false, false-true and true-true values 
print(cm)

[[45  8]
 [14 76]]


# Conclusion:
We can conclude that the model has predicted the cancer to be benign or malignant with an accuracy of more than 85%(Cannot give a single value as the value depends on the data spliting).

The optimum threshold value is found to be 26,27 or 28 depending on the data split.

The accuracy values for train and test data are nearly same so the model can be said as a error free and stable model.